server/server/src/test/java/org/infinispan/server/server/ClusteredServerTest.java

```diff
@@ -0,0 +1,44 @@
+package org.infinispan.server.server;
+
+import static org.junit.Assert.assertEquals;
+
+import org.infinispan.client.hotrod.RemoteCache;
+import org.infinispan.client.hotrod.RemoteCacheManager;
+import org.infinispan.server.server.test.ServerTestMethodConfiguration;
+import org.infinispan.server.server.test.ServerTestMethodRule;
+import org.infinispan.server.server.test.ServerTestConfiguration;
+import org.infinispan.server.server.test.ServerTestRule;
+import org.junit.ClassRule;
+import org.junit.Rule;
+import org.junit.Test;
+
+/**
+ * @author Tristan Tarrant &lt;tristan@infinispan.org&gt;
+ * @since 10.0
+ **/
+@ServerTestConfiguration
```

Comment:
Testing wise, I'd like to see us use suites. E.g. a cluster suite would start a cluster of 2 nodes, and then it executes all the tests and combinations against that cluster. The granularity of suites would need to be discussed, but I'd like to avoid each test start/stop servers. This hugely speeds up testing but the downside is that you have to make sure tests don't leak anything. I've used this approach for Node.js client testsuite and IIRC, it's inspired by what @gustavonalle did with the Spark module?

Reply:
Related to this, what is the lifecycle of `@ClassRule` and `@Rule`? I'm not sure they're designed for my suggested approach?

Reply:
Yes, the intention is to use the Class Rule in conjuction with the Suite runner to give you the same.

Reply:
It'd be nice to extract test classes like  ```Cluster``` and ```Node``` so that it's decoupled and can be used anywhere regardless of annotations or test frameworks. It can then be invoked by @Rule.

I have those classes separated in the spark connector suite, and it allows me to use it anywhere, example:

https://github.com/gustavonalle/infinispan-cluster-test/blob/master/src/test/java/org/infinispan/ClusterTest.java

Another idea, is to have two variants of the Cluster classes, one for embedded clusters and other for servers in separate JVMs.

Reply:
👍

Reply:
https://issues.jboss.org/browse/ISPN-9945